### PR TITLE
Add Custom Render Plugin Example

### PR DESCRIPTION
**Pull request by JShep1**
The last piece to getting the custom render engine issue (https://github.com/ignitionrobotics/ign-rendering/issues/100) resolved.

This example simply adds a mock render engine plugin that the user can load with `--render-engine HelloWorldPlugin`, it won't display anything to the scene as there is actually no implementation, this just serves as a template for any future user desiring to write their own render engine plugin.

Signed-off-by: John Shepherd <john@openrobotics.org>

Depends on https://github.com/ignitionrobotics/ign-gazebo/pull/373

### CODE REVIEW

**Comment by iche033** [comment]
The [changes here](https://github.com/ignitionrobotics/ign-rendering/compare/jshep1/add_custom_plugin_example...ian/add_custom_plugin_example?expand=1) allows the plugin to load in the ign-rendering side. It's now just failing in RenderUtil on the in-gazebo side since there are several places that assume a non-null scene..

**Comment by JShep1** [comment]
> The [changes here](https://github.com/ignitionrobotics/ign-rendering/compare/jshep1/add_custom_plugin_example...ian/add_custom_plugin_example?expand=1) allows the plugin to load in the ign-rendering side. It's now just failing in RenderUtil on the in-gazebo side since there are several places that assume a non-null scene..

Can you make the PR?  Mainly so I can approve.  Also, I made an additional PR for ign-gazebo [here](https://github.com/ignitionrobotics/ign-gazebo/pull/394) adding additional null checks.  I'm still getting a seg fault when attempting to run the mock render engine plugin, and `gdb` actually isn't helpful at all so I'm trying to figure out where exactly it's occurring.

**Comment by iche033** [comment]
created pull request #156 containing the null checks, and your changes :)

**Comment by JShep1** [comment]
This is ready for a review whenever you have time. @iche033

File: examples/hello_world_plugin/HelloWorldPlugin.cc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    /// \brief Singelton setup.
+    private: friend class ignition::common::SingletonT<HelloWorldRenderEngine>;
+  };
+
+  class HelloWorldPlugin :
+    public ignition::rendering::RenderEnginePlugin
+  {
+    /// \brief Get the name of the render engine loaded by this plugin.
+    /// \return Name of render engine
+    public: std::string Name() const
```
Comment (iche033) [inline examples/hello_world_plugin/HelloWorldPlugin.cc:83]: I think this and the function below can be marked `override`

File: examples/hello_world_plugin/HelloWorldPlugin.cc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    // Documentation Inherited.
+    public: virtual std::string Name() const override
+    {
+      return "HelloWorldPlugin";
+    }
+
+    // Documentation Inherited.
+    protected: virtual bool LoadImpl(const std::map<std::string, std::string> &_params) override
```
Comment (iche033) [inline examples/hello_world_plugin/HelloWorldPlugin.cc:45]: `sh tools/code_check.sh` caught a few style errors. Can you wrap the lines to 80 char?

File: examples/hello_world_plugin/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# Engine plugin
+
+This example shows how to create a plugin that integrates a rendering engine with
+Ignition Rendering.
```
Comment (iche033) [inline examples/hello_world_plugin/README.md:4]: Can you add a sentence that says this demo also shows how to load this plugin with Ignition Gazebo?

File: examples/hello_world_plugin/HelloWorldPlugin.cc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+#include <ignition/plugin/Register.hh>
+
+#include "ignition/rendering/RenderEnginePlugin.hh"
+#include "ignition/rendering/base/BaseRenderEngine.hh"
+#include "ignition/rendering/base/BaseRenderTypes.hh"
+
+namespace mock
+{
+  class HelloWorldRenderEngine :
```
Comment (iche033) [inline examples/hello_world_plugin/HelloWorldPlugin.cc:28]: Can you add a `\brief` doxygen comment for this and the `HelloWorldPlugin` class?

**Review by iche033 — COMMENTED** [review]
Nice it's working for me! I am able load the hellow world plugin following the intructions in the README.md. Just some minor style and doc comments

File: examples/hello_world_plugin/HelloWorldPlugin.cc (hunk 2 above)
Comment (JShep1) [reply examples/hello_world_plugin/HelloWorldPlugin.cc]: 0dd768276f5742bd222ca33340f1035a413542aa

File: examples/hello_world_plugin/README.md (hunk 3 above)
Comment (JShep1) [reply examples/hello_world_plugin/README.md]: 0dd768276f5742bd222ca33340f1035a413542aa

File: examples/hello_world_plugin/HelloWorldPlugin.cc (hunk 4 above)
Comment (JShep1) [reply examples/hello_world_plugin/HelloWorldPlugin.cc:29]: 0dd768276f5742bd222ca33340f1035a413542aa

File: examples/hello_world_plugin/HelloWorldPlugin.cc (hunk 1 above)
Comment (JShep1) [reply examples/hello_world_plugin/HelloWorldPlugin.cc]: 0dd768276f5742bd222ca33340f1035a413542aa

**Review by iche033 — APPROVED** [review]

File: src/base/BaseRenderEngine.cc [hunk 5]
```diff
@@ -94,72 +94,108 @@ bool BaseRenderEngine::IsEnabled() const
...
+  return ScenePtr();
 }
 
 //////////////////////////////////////////////////
 void BaseRenderEngine::DestroyScene(ScenePtr _scene)
 {
+  auto scenes = this->Scenes();
+  if (!scenes)
+    return;
   this->Scenes()->Destroy(_scene);
```
Comment (chapulina) [inline src/base/BaseRenderEngine.cc:163]: Nit: you could use `scenes` on all functions from this one to the bottom.

**Review by chapulina — APPROVED** [review]
:+1:

File: src/base/BaseRenderEngine.cc (hunk 5 above)
Comment (JShep1) [reply src/base/BaseRenderEngine.cc]: https://github.com/ignitionrobotics/ign-rendering/pull/154/commits/35a3c5b37c26d5cc97a95872cf30dcbf012d9b31

**Pull request merged**